File: csharp/robot-name/RobotName.cs
using System;

namespace robot
{
	public class Robot
	{
		private static Random _random = new Random ();

		public string Name { get; private set; }

		public Robot ()
		{
			Reset ();
		}

		public void Reset ()
		{
			Name = GenerateName ();
		}

		private string GenerateNumber ()
		{
			return _random.Next (1000).ToString ("000");
		}

		private char GenerateCharacter ()
		{
			return (char)('A' + _random.Next (26));
		}

		private string GenerateName()
		{
			return String.Format ("{0}{1}{2}", GenerateCharacter (), GenerateCharacter (), GenerateNumber ());
		}
	}
}

File: ruby/robot-name/robot_name.rb
class Robot
  VERSION = 1

  attr_reader :name

  @taken_names = []

  def name
    @name ||= generate_name
  end

  def reset
    @name = nil
  end

  def self.taken_names
    @taken_names
  end

  private

  def generate_name
    name = generate_letters + generate_number
    if self.class.taken_names.include? name
      name = generate_name
    end
    self.class.taken_names << name
    name
  end

  def generate_letters
    [*('A'..'Z')].sample(2).join
  end

  def generate_number
    rand.to_s[2..4]
  end
end